### PR TITLE
Align ZeroShotClassificationTranslator with Hugging Face behavior

### PR DESCRIPTION
**Pull request by raphaeldelio**
This refactor improves the ZeroShotClassificationTranslator implementation to better reflect the behavior of Hugging Face’s zero-shot-classification pipeline in Python. It replaces several hardcoded assumptions with dynamic logic that adapts based on model configuration.

### Motivation

The translator was optimized to work with the "djl://ai.djl.huggingface.pytorch/facebook/bart-large-mnli" model. When a different model was used, it would either output wrong predictions or throw errors altogether. 

Some models on Hugging Face define their label mapping in a non-standard order. Without reading and honoring label2id, the results returned by the translator may be incorrect or misleading. The previous implementation also failed on models that require token type IDs. These issues are addressed in this refactor.

The new implementation closely mirrors the logic used in the Hugging Face Transformers ZeroShotClassificationPipeline, improving correctness and interoperability with a wider range of models.

⸻

I created a repo to compare the outputs of the new refactored translator with the outputs of Python's Transformers library: https://github.com/raphaeldelio/deep-java-library-zero-shot-classification/

I compared the outputs of four different models and they were all consistent:

MoritzLaurer/DeBERTa-v3-large-mnli-fever-anli-ling-wanli
Facebook/bart-large-mnli    
MoritzLaurer/bge-m3-zeroshot-v2.0   
tasksource/ModernBERT-base-nli

### Summary of changes

1. Dynamic entailment and contradiction label IDs
	•	Previous implementation assumed fixed label indices (entailment = 2, contradiction = 0).
	•	This change parses the label2id map from config.json, if available, to dynamically determine the correct label positions.
	•	Fallback values are retained for backward compatibility.
	•	The builder now allows manually overriding these IDs if needed.

2. Support for token_type_ids
	•	Added inference of withTokenType based on type_vocab_size and model_type in the model config.
	•	Common transformer families like BERT, ALBERT, and DeBERTa use token type embeddings and require this flag.
	•	The builder supports overriding withTokenType explicitly.

3. Updated multi-label and single-label scoring logic
	•	Multi-label classification now applies a softmax over the [contradiction, entailment] logits for each candidate and returns the entailment probability. This matches Hugging Face’s logic.
	•	Single-label classification uses only the entailment logits, then applies softmax across all candidates to normalize scores. This avoids including irrelevant logits from the contradiction class.
	•	Sorting and label-score pairing were updated to correctly reflect probabilities and preserve candidate order.

4. Config-aware translator preparation
	•	During prepare(), the model’s config.json is parsed to extract relevant metadata (label mappings, type vocab size, model type).
	•	Failures in config parsing are logged but do not break inference.

5. Extended builder API

	•	The builder now supports:
	•	withTokenType(boolean)
	•	int32(boolean)
	•	entailmentId(int)
	•	contradictionId(int)
	•	This allows downstream users to override behavior explicitly when needed.

### Impact

	•	Existing behavior is preserved for models that follow standard label conventions.
	•	Compatibility is improved for models with custom label IDs and token type requirements.
	•	The new translator is more robust and predictable

### CODE REVIEW

**Review by frankfliu — COMMENTED** [review]
Thank you for your contribution.

Please follow our development guide line: https://docs.djl.ai/master/docs/development/development_guideline.html#development-guideline
sepcifically to fix coding style and pass the build checks:
1. format the java code with: `./gradlew fJ`
2. pass all checks: `./gradlew check`

File: extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java [hunk 1]
```diff
@@ -31,40 +31,131 @@
...
 
-    ZeroShotClassificationTranslator(HuggingFaceTokenizer tokenizer, boolean int32) {
+    ZeroShotClassificationTranslator(HuggingFaceTokenizer tokenizer, boolean tokenTypeId, boolean int32) {
         this.tokenizer = tokenizer;
+        this.tokenTypeId = tokenTypeId;
         this.int32 = int32;
     }
 
-    /** {@inheritDoc} */
+    // Constructor to allow passing entailment/contradiction IDs if known
```
[the remark this answers is not in the document]
Comment (raphaeldelio) [reply extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java]: Unnecessary comment removed

**Comment by raphaeldelio** [comment]
> Thank you for your contribution.
> 
> Please follow our development guide line: https://docs.djl.ai/master/docs/development/development_guideline.html#development-guideline sepcifically to fix coding style and pass the build checks:
> 
> 1. format the java code with: `./gradlew fJ`
> 2. pass all checks: `./gradlew check`

Thank you for referencing the guidelines. I'd missed it. 

I ran the formatting commands, fixed the javadocs, and applied your suggestions.

File: extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java [hunk 2]
```diff
@@ -29,32 +29,135 @@
...
+        this.contradictionId = contradictionId;
+    }
+
     /** {@inheritDoc} */
     @Override
     public void prepare(TranslatorContext ctx) throws IOException, ModelException {
         Model model = ctx.getModel();
         predictor = model.newPredictor(new NoopTranslator(null));
         ctx.getPredictorManager().attachInternal(UUID.randomUUID().toString(), predictor);
+
```
Comment (frankfliu) [inline extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java:91]: The code can be simplified a bit:

```
        Path configFile = model.getModelPath().resolve("config.json");
        if (!Files.isRegularFile(configFile)) {
            return;
        }
        try (Reader reader = Files.newBufferedReader(configFile)) {
            JsonObject config = JsonUtils.GSON.fromJson(reader, JsonObject.class);
            if (config.has("label2id")) {
                JsonObject label2Id = config.getAsJsonObject("label2id");
                if (label2Id.size() == 3) {
                    // only for 3-class
                    for (Map.Entry<String, JsonElement> entry : label2Id.entrySet()) {
                        String key = entry.getKey().toLowerCase(Locale.ROOT);
                        int value = entry.getValue().getAsInt();
                        if (key.startsWith("entail")) {
                            entailmentId = value;
                        } else if (key.startsWith("contra")) {
                            contradictionId = value;
                        }
                    }
                }
            }

            boolean inferredWithTokenType = false; // Default assumption
            if (config.has("type_vocab_size")) {
                JsonElement typeVocabSizeObj = config.get("type_vocab_size");
                if (typeVocabSizeObj.isJsonPrimitive()) {
                    int typeVocabSize = typeVocabSizeObj.getAsInt();
                    if (typeVocabSize > 1) {
                        inferredWithTokenType = true;
                    }
                }
            }

            if (!inferredWithTokenType && config.has("model_type")) {
                String modelType = config.get("model_type").getAsString();
                if ("bert".equals(modelType)
                        || "albert".equals(modelType)
                        || "xlnet".equals(modelType)
                        || "deberta".startsWith(modelType)) {
                    inferredWithTokenType = true;
                }
            }

            tokenTypeId = inferredWithTokenType;
        } catch (IOException | JsonParseException e) {
            logger.error("Failed to read or parse config.json for label2id", e);
        }
```

File: extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java [hunk 3]
```diff
@@ -82,32 +185,75 @@ public ZeroShotClassificationOutput processOutput(TranslatorContext ctx, NDList
...
+                        combinedLogits.get(
+                                new NDIndex(
+                                        ":, {}",
+                                        manager.create(new int[] {contradictionId, entailmentId})));
+                NDArray scoresProbs = entailContrLogits.softmax(1);
+                entailmentScores = scoresProbs.get(":, 1");
+            }
+
+            float[] floatScores = entailmentScores.toFloatArray();
+            double[] tempScores = new double[floatScores.length];
```
Comment (frankfliu) [inline extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java:216]: Use DJL `Pair` can make the code a bit shorter:

```
            List<Pair<Double, String>> pairs = new ArrayList<>();
            for (int i = 0; i < floatScores.length; i++) {
                Pair<Double, String> pair = new Pair<>((double) floatScores[i], candidates[i]);
                pairs.add(pair);
            }
            pairs.sort(
                    Comparator.comparingDouble((Pair<Double, String> e) -> e.getKey()).reversed());
```

File: extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java [hunk 4]
```diff
@@ -136,8 +282,9 @@ public static Builder builder(HuggingFaceTokenizer tokenizer) {
      * @param arguments the models' arguments
      * @return a new builder
      */
-    public static Builder builder(HuggingFaceTokenizer tokenizer, Map<String, ?> arguments) {
-        Builder builder = builder(tokenizer);
+    public static ZeroShotClassificationTranslator.Builder builder(
```
Comment (frankfliu) [inline extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java:285]:
```suggestion
    public static Builder builder(
```

File: extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java [hunk 5]
```diff
@@ -136,8 +282,9 @@ public static Builder builder(HuggingFaceTokenizer tokenizer) {
...
+            HuggingFaceTokenizer tokenizer, Map<String, ?> arguments) {
+        ZeroShotClassificationTranslator.Builder builder = builder(tokenizer);
```
Comment (frankfliu) [inline extensions/tokenizers/src/main/java/ai/djl/huggingface/translator/ZeroShotClassificationTranslator.java:287]:
```suggestion
        Builder builder = builder(tokenizer);
```

**Review by frankfliu — APPROVED** [review]
This is great

**Review by xyang16 — APPROVED** [review]
